### ISSUE TITLE
Azure Mobile Services returns MobileServicePushFailedException after calling PullAsync()

### ISSUE THREAD

**Issue by gonzobrains** [Azure/azure-mobile-apps-net-client#437]
I have created a Xamarin.Android application that uses Azure Mobile Services. I am receiving a MobileServicePushFailedException after calling PullAsync().

After inspecting the exception, I see that my client is trying to send Update operations instead of Insert operations for new rows created on the client that do not yet exist on the server. The server sees an update come through for a non-existent row and returns a 404 error.

Why is this happening and how can I fix it?

**Comment by MatkovIvan** [comment]
Hi @gonzobrains 

Thanks for getting in touch!
You are talking about an exception on pull operation, but then about sending updates in the exception details. But pull operation is (from [Offline Data Sync doc](https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-offline-data-sync#how-sync-works)):
> **Pull**: Pull is performed on a per-table basis and can be customized with a query to retrieve only a subset of the server data. The Azure Mobile client SDKs then insert the resulting data into the local store.

Currently, description is very confusing and unclear. Please provide the exception details and reproduction steps.

Best,
Ivan

**Comment by gonzobrains** [comment]
It is my understanding that an implicit "Push" occurs when PullAsync() is called.  Is this not correct?  In any case, I have a similar problem when calling PushAsync() as well.  Basically the MobileServicePushFailedException contains information that seems to indicate that Update operations are occurring for new rows instead of Inserts.  I think this is happening due to a timeout occurring on the first insert. 

Here is what I see in the first PushResult's RawResult:

"The operation failed with the following error: 'The wait operation timed out'"

And since the rest of the updates are dependent on this first row being inserted, I then get several of the following in separate PushResult objects:

"The INSERT statement conflicted with the FOREIGN KEY constraint"

How can I stop this timeout from occurring?  At the very least, can I make the sync fail immediately after the first timeout instead of allowing the sync to continue with further failures?

**Comment by MatkovIvan** [comment]
> an implicit "Push" occurs when PullAsync() is called.

That's true

> FOREIGN KEY

Please note, Azure Mobile Apps does not support relationships.
Related topics:
Azure/azure-mobile-apps#6
https://shellmonger.com/2016/05/27/30-days-of-zumo-v2-azure-mobile-apps-day-26-relationship-advice/

From this, I don't think that timeout is the root cause. BTW Is it occurs on sample app or only on your main application?

**Comment by gonzobrains** [comment]
I have not tried the sample yet.  Could you provide me a link to the one to
which you arE referring?

Since I inherited this project, I am not sure if the back-end is using
Node.js or ASP.NET as I only work on the mobile app side.  How can I
check?  That being said, I am pretty sure the developer who works on the
server side stuff is developing a Web app in C# to accept my mobile app's
calls.

**Comment by MatkovIvan** [comment]
You can use [quickstart app from azure portal](https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-xamarin-forms-get-started#download-and-run-the-xamarinforms-solution) as sample app.
The easiest way is to ask the backend developer.

Did this issue appear after some changes (client or server)? Has syncing worked in your project before?
What errors do you see in the server-side logs?

**Comment by gonzobrains** [comment]
Ivan,

I forgot to ask if you can suggest any retry logic I should use.  Is there
also a way to make PullAsync() and PushAsync() fail immediately on the
first error so I can retry?  If this timeout would return right away I
could try again instead of having the sync continue attempting to insert
records that are dependent on rows in other tables being inserted first.

Either that or increase the timeout.

Lastly, the errors seem to occur on just one table.  Often times
the MobileServicePushFailedException thrown contains rows with Handled set
to false but no Result, Status, or anything else to indicate why it failed.

Why is this?  How can I figure out what is causing the problem?

Thanks again,
Jeff

**Comment by MatkovIvan** [comment]
> So that should work with relationships, correct?

Yep, but only as described in articles by links above.

> "The operation failed with the following error: 'The wait operation timed out'"

I believe that's **NOT** a HTTP timeout: it's the string returned by server. So, it's an error in some internal server operation (connect to database or so).

> fail immediately on the first error so I can retry?

The network error aborts the process out of the box.

> or increase the timeout.

There is no public API for it. #414

> How can I figure out what is causing the problem?

Try to checkout `InnerException`

Best,
Ivan

**Comment by gonzobrains** [comment]
Hi,

What exactly do you mean by "The network error aborts the process out of
the box?"

I have witnessed immediate exceptions thrown when a network error occurs,
but the timeout I mention is actually in
the MobileServicePushFailedException.  Does that mean it's an internal
server error as you mentioned before?  Do you have an example that shows
the best type of error handling I should implement in this case (and for
any case, for that matter)?

Lastly, can I use my MSDN support options for paid support should I be
unable to solve this issue via this forum?

Thanks again,
Jeff

**Comment by MatkovIvan** [comment]
The current logic of the SDK is if network error happens - fail immediately and don't try subsequent operation. Unfortunately, HTTP timeout throws `OperationCanceledException` and it isn't treat as network error for now. I've made a PR to fix this behavior (#458).
But in your case the exception has `RawResult` field - it contains the string **returned by server**, so it isn't an HTTP timeout.

You can deal with errors/conflicts or just monitor each operation in particular via implementing `IMobileServiceSyncHandler`. You can also abort the push by some condition (inside `ExecuteTableOperationAsync`) if you want.

**Comment by dhei** [comment]
Hey @gonzobrains,
Let us know if you still have issue after trying @MatkovIvan's suggestions.
Please feel free to create Azure support ticket if you prefer to do so.

**Comment by gonzobrains** [comment]
How do I create an Azure support ticket?  Through the portal?

**Comment by guperrot** [comment]
You can create a support ticket on the Azure portal from your app service resource.

**Comment by gonzobrains** [comment]
Hi Ivan,

I have been experimenting with the "Coffee Cups" app:

https://blog.xamarin.com/getting-started-azure-mobile-apps-easy-tables/

Is this your reference app or do you suggest something else to try?  I was
able to set up the back-end and the app uploads coffee updates.  This is a
fairly simple app, though.  I'm not sure how to go from here to determine
what's causing my app to fail.  My app has a lot more tables and tables
that have relationships.

Since the PushResult objects are not returning any helpful information I
can't diagnose what the actual problem is.

What should I try next?

Thanks again,
Jeff

**Comment by MatkovIvan** [comment]
As I wrote before:
> I believe that's **NOT** a HTTP timeout: it's the string returned by server. So, it's an error in some internal server operation (connect to database or so).

> in your case the exception has `RawResult` field - it contains the string **returned by server**, so it isn't an HTTP timeout.

So, it looks like a server-side issue. I guess you should find out why the server sends this error to the client side and what operation has a timeout.

Please let me know if you have any details that might disprove my assumption.

**Comment by gonzobrains** [comment]
Hi Ivan,

I haven't seen the timeout problems anymore.  But I do see the problem
where the PushResult objects don't provide any additional info.  I will
research further.  Our backend person has returned from vacation, so
hopefully I can investigate further on the server side.

However, I do recall that our original sync problems stemmed from database
UPDATEs being sent instead of INSERTs for new records.  Could that
originate from the client-side or would that be a problem caused by the
ASP.NET side?

Lastly, I read that article on relationships.  The problematic table in our
project refers to several foreign keys and several other tables refer to
it.  Is this supported?  If not, should we expect a specific error to be
raised or would the symptoms we currently see be caused by our specific
relationship setup not being supported?  I'd like to try to determine if we
ultimately have to revise our database design to make this work correctly.
I don't think the original developer was aware of this limitation.  If
there are any specific steps I can take to confirm this please let me
know.  I have SSMS and can look through the database diagram, etc. to see
if there is anything that may not be compatible with the Azure Mobile
Services library.

Thanks again,
Jeff

**Comment by gonzobrains** [comment]
@dhei I will open a developer support ticket through the Azure portal shortly.

@MatkovIvan After working with the backend developer, it looks as though we not only see PATCH requests when we should see POST requests, but we also see POSTs come through when we should see PATCH requests.  We see this by remotely attaching to the backend and stepping through the calls with a debugger.

When we try to sync again, we start to get conflicts.  I'm not sure why this would happen when this is all new data being sent to the server.  I confirmed with the backend developer we are not using any many-to-many relationships.

I will try to create a sample that replicates this that I can share with you.

**Comment by MatkovIvan** [comment]
> Could that originate from the client-side or would that be a problem caused by the ASP.NET side?

I guess it's kind of integration problem where an initial error happens on one side but doesn't handle correctly by other side. Unfortunately, I cannot say more precisely due to lack of reproduce.

> Is this supported?

You said it worked in the past. Nothing was supposed to change.

> I will try to create a sample that replicates this that I can share with you.

That would be great.

**Comment by serri588** [comment]
Hello!

So I'm the backend dev on the project, I also have the joy of inheriting it- learning as I go here. I've been reading through the thread and the article you linked. I think I'm getting some insight into why we are having issues. We can still work on the sample.

To use the TodoItem & Tags example, the API is expecting the send and receive the tags as an array attached to the TodoItem, not as a separate entity. 

What we are trying to do:
Insert TodoItem1  {properties}
Insert Tag1 {properties}
Insert Tag2 {properties}

What the client-side library thinks we are doing:
Insert TodoItem1 {properties, Tags: [Tag1, Tag2]}
Update Tag1 {properties}
Update Tag2 {properties}

So what I think is happening is that the client needs to send me "Insert TodoItem1 {properties, Tags: [Tag1, Tag2]}" and I need to update that endpoint to insert both the TodoItem and the corresponding Tags.

I'm not sure if that would completely solve our issues, but it sounds like that would be a good start.

**Comment by gonzobrains** [comment]
Serri588,

Hi and thanks for joining the conversation.  Where can I get the "Todo"
sample you mentioned?

Jeff

**Comment by gonzobrains** [comment]
I thought it would be worth trying to download the source and build.  The
README file says all I have to do is update nuget and hit F6.

That's not working for me.  I get a ton of build errors.  Plus I see the
yellow warning triangle on some of the dependencies.

What am I missing here?

I just want to build for Xamarin.Android, include it in my project, and
step through the library's code to see if I can find any insight to the
cause of our sync problems.

**Comment by MatkovIvan** [comment]
To build the SDK from source please make sure that:
1. You downloaded git submodules
2. You are on PC with VS 2017 and [connected Mac](https://docs.microsoft.com/en-us/xamarin/ios/get-started/installation/windows/connecting-to-mac/). This is required because the SDK has platform depended code for UWP (only Windows) and iOS (only macOS). If you unable to get this configuration you can remove unnecessary for your case platforms from [this line](https://github.com/Azure/azure-mobile-apps-net-client/blob/master/src/Microsoft.Azure.Mobile.Client/Microsoft.Azure.Mobile.Client.csproj#L5).

**Comment by gonzobrains** [comment]
@MatkovIvan 

1.  I downloaded from git and not from a .zip file so all the submodules should be there, correct?  How can I verify?

2.  I am only set up for Android at the moment so I don't need UWP or a Mac build agent.  In Visual Studio's Solution Explorer I see MonoAndroid 7.0 and Xamarin.iOS 1.0 have yellow warning triangles next to them.  In the .csproj file I see the line you mentioned is for "TargetFrameworks."  My .csproj file does not have that.  Mine only has a "TargetFrameworkVersion."  Is that because my project is only building a Xamarin.Android apk?  How do I get the warning triangle to go away on MonoAndroid 7.0 if I need to build this for my Android project?  Is this different from my project because it is a class library?

3.  If I want to include this library in my solution instead of from nuget so I can debug it which parts of your solution do I need to include?

4.  Can you provide step-by-step instructions for building the library for Android, testing it with your provided test framework, and then including the library in one's own project in place of the pre-made NuGet version so that we can not only test/debug it but potentially maintain it ourselves in the future?  Do I need to build and include the SQLiteStore project as well?  

I can see from your README file that your team is not investigating any new feature work.  If this is the case, what do you foresee will eventually replace this library?  When will the next bug-fix version be released?

Please advise.

Regards,
Jeff

**Comment by MatkovIvan** [comment]
Hi Jeff,

1. Check out the content in the `src/External/Xamarin.Auth` - this folder shouldn't be empty.
2. Yes, there are differences between application and library projects. Feel free to edit `TargetFrameworks` tag and remove extra entries.
3. `Microsoft.Azure.Mobile.Client.csproj` and `Microsoft.Azure.Mobile.Client.SQLiteStore.csproj`
4. You don't need the our tests applications for using the SDK project in the your solution. To use in in your project:

    * Open you solution;
    * Remove nuget dependency from your project;
    * Add `Microsoft.Azure.Mobile.Client.csproj` and `Microsoft.Azure.Mobile.Client.SQLiteStore.csproj` as existing projects in your solution;
    * Remove extra targets from SDK. Open `.csproj` in any text editor and modify `TargetFrameworks` tag: `<TargetFrameworks>MonoAndroid70</TargetFrameworks>`;
    * Add SDK project as project dependency into your application;<br/><br/>
    
    > potentially maintain it ourselves in the future? 

    You're always welcome to make contributions via pull requests in this repository 😃 

    > Do I need to build and include the SQLiteStore project as well?
    
    Yes, to avoid potential conflicts


> what do you foresee will eventually replace this library? When will the next bug-fix version be released?

@elamalani could you please answer to these questions?

Thanks,
Ivan

**Comment by gonzobrains** [comment]
@Ivan Thanks.  
1.  The subdir was empty.  I use sourcetree, so I had to open a terminal and run the gitsubmodule commands:

git submodule init
git submodule update

Unfortunately, it still doesn't build.  There is no obj subdir so it can't find a lot of files.  I think it has to do with 2).

2.  I removed the extra entries (i.e., iOS, UWP), but I don't want to remove MonoAndroid 7.0 because I am targeting Android!  How do I get the yellow triangles to go away?  I don't see any way to either add MonoAndroid 7.0 to the project or update it to 8.0.  In my project I see a Xamarin.Android references, not MonoAndroid dependencies.  What's the difference and how do I fix it?

The project claims to have been built, but it didn't.  Please see the attached screen snippet and build output.
[MicrosoftAzureMobileClient_fail_build_output.txt](https://github.com/Azure/azure-mobile-apps-net-client/files/2416915/MicrosoftAzureMobileClient_fail_build_output.txt)

![yellow_triangles](https://user-images.githubusercontent.com/364198/46040588-bf539a00-c0de-11e8-8268-d2b477eb5da6.PNG)

**Comment by gonzobrains** [comment]
@dhei Azure Portal support is not sure they can be of any assistance.  Do you know of what Microsoft resource I can contact for custom consulting beyond what can be handled here?  Thanks.

**Comment by gonzobrains** [comment]
I am looking in the Microsoft.Azure.Mobile.Client.SQLiteStore project and I also see yellow triangles on .NETFramework 4.5, .NETPortable 0.0, and .NETStandard 1.4

**Comment by gonzobrains** [comment]
Okay, it looks like updating Visual Studio improved things.  I now see Mono.Android instead of Mono.Android 7.0 and no yellow triangles. in the Microsoft.Azure.Mobile.Client project.  However, I still see yellow triangles in the Microsoft.Azure.Mobile.Client.SQLiteStore project for some assemblies (see photo).

![assemblies_yellow_triangles](https://user-images.githubusercontent.com/364198/46094550-3d1fb000-c188-11e8-8d8b-48ac6326db1d.PNG)

The projects do build now.  I had to remove all the existing TargetFrameworks for both .csproj files and replace them with MonoAndroid70.  The SQLiteStore project didn't have MonoAndroid70 before.  Any reason for that?

Is there a better way to manage these TargetFramework settings?  There should be an easier way to fix these "yellow triangle" problems as well.

I will report back if these DLLs work as valid replacements for the NuGet ones before attempting to add the projects to my solution for debugging.

**Comment by bmourat** [comment]
Hey @gonzobrains 
The SQLiteStore project doesn't have MonoAndroid70 target cause it doesn't contain platform dependent code. As for the yellow triangles, they can be caused by multiple reasons. Not sure what is the case here, but I think you can ignore them. I've analyzed your build output and it indeed seems like the project was built successfully, just a couple of warnings. Why do you think build failed?

**Comment by gonzobrains** [comment]
@bmourat 

I thought it failed as I didn't see any output folders yesterday.  I may have been looking in the wrong place for the obj/bin folders.  I see them here today.  I had to unload all the test-related projects to not see any other errors.  I would eventually like to try them out, especially if we have to start fixing bugs in the project ourselves locally.

What should I set the SQLiteStore target to be then?  Is editing the .csproj file directly the only way to change this?  I eventually want to target Android 8.0 as that's what my project targets.  

What's the difference between "Mono.Android" and "Xamarin.Android" anyway?
